ratis-common/src/main/java/org/apache/ratis/util/Daemon.java

```diff
@@ -17,24 +17,59 @@
  */
 package org.apache.ratis.util;
 
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Objects;
+
 public class Daemon extends Thread {
+  static final Logger LOG = LoggerFactory.getLogger(Daemon.class);
+  public static final Thread.UncaughtExceptionHandler LOG_EXCEPTION =
+      (t, e) -> LOG.error(t.getName() + " threw an uncaught exception", e);
+
   {
     setDaemon(true);
   }
 
-  /** Construct a daemon thread. */
-  public Daemon() {
-    super();
+  /** Construct a daemon thread with flexible arguments. */
+  protected Daemon(Builder builder) {
+    super(builder.runnable);
+    setName(builder.name);
+    if (builder.uncaughtExceptionHandler != null) {
+      setUncaughtExceptionHandler(builder.uncaughtExceptionHandler);
+    }
   }
 
-  /** Construct a daemon thread with the given runnable. */
-  public Daemon(Runnable runnable) {
-    this(runnable, runnable.toString());
+  /** @return a {@link Builder}. */
+  // TODO(jiacheng): For exceptions that should crash the server, apply a correct UncaughtExceptionHandler
+  public static Builder newBuilder() {
+    return new Builder();
   }
 
-  /** Construct a daemon thread with the given runnable. */
-  public Daemon(Runnable runnable, String name) {
-    super(runnable);
-    this.setName(name);
+  public static class Builder {
+    private String name;
+    private Runnable runnable;
+    // By default, uncaught exceptions are just logged without further actions
+    private UncaughtExceptionHandler uncaughtExceptionHandler = LOG_EXCEPTION;
```

Comment:
Use null for default.

Reply:
As mentioned above, I figure logging as the default action doesn't break existing behavior and leaves a way for users to identify issues so they can be fixed later.